tests/ixia/pfcwd/files/pfcwd_runtime_traffic_helper.py

```diff
@@ -0,0 +1,224 @@
+            packet=[
+                Header(choice=EthernetHeader(pfc_queue=pfc_queue)),
+                Header(choice=Ipv4Header(priority=ip_prio))
+            ],
+            size=Size(data_pkt_size),
+            rate=Rate('line', data_flow_rate_percent),
+            duration=Duration(FixedSeconds(seconds=data_flow_dur_sec))
+        )
+
+        result.append(data_flow)
+
+    return result
+
+def __run_traffic(api, config, duthost, all_flow_names, pfcwd_start_delay_sec, exp_dur_sec):
+    """
+    Start traffic at time 0 and enable PFC watchdog at pfcwd_start_delay_sec
+
+    Args:
+        api (obj): IXIA session
+        config (obj): experiment config (testbed config + flow config)
+        duthost (Ansible host instance): device under test
+        all_flow_names (list): list of names of all the flows
+        pfcwd_start_delay_sec (int): PFC watchdog start delay in second
+        exp_dur_sec (int): experiment duration in second
+
+    Returns:
+        per-flow statistics (list)
+    """
+
+    api.set_state(State(ConfigState(config=config, state='set')))
+    api.set_state(State(FlowTransmitState(state='start')))
+
+    time.sleep(pfcwd_start_delay_sec)
+    start_pfcwd(duthost)
+    time.sleep(exp_dur_sec - pfcwd_start_delay_sec)
+
+    attempts = 0
+    max_attempts = 20
+
+    while attempts < max_attempts:
+        rows = api.get_flow_results(FlowRequest(flow_names=all_flow_names))
+        """ If all the flows have stopped """
+        transmit_states = [row['transmit'] for row in rows]
+        if len(rows) == len(all_flow_names) and\
+           list(set(transmit_states)) == ['stopped']:
+            time.sleep(IXIA_POLL_DELAY_SEC)
+            break
+        else:
+            time.sleep(1)
+            attempts += 1
+
+    """ Dump per-flow statistics """
+    rows = api.get_flow_results(FlowRequest(flow_names=all_flow_names))
+    api.set_state(State(FlowTransmitState(state='stop')))
+
+    return rows
+
+def __verify_results(rows, speed_gbps, data_flow_dur_sec, data_pkt_size, tolerance):
+    """
+    Verify if we get expected experiment results
+
+    Args:
+        rows (list): per-flow statistics
+        speed_gbps (int): link speed in Gbps
+        data_flow_dur_sec (int): duration of data flows in second
+        data_pkt_size (int): size of data packets in byte
+        tolerance (float): maximum allowable deviation
+
+    Returns:
+        N/A
+    """
+    data_flow_rate_percent = int(100 / len(rows))
+
+    for row in rows:
+        flow_name = row['name']
+        tx_frames = row['frames_tx']
+        rx_frames = row['frames_rx']
+
+        pytest_assert(tx_frames == rx_frames, "{} packets of {} are dropped".\
+                      format(tx_frames-rx_frames, flow_name))
+
+        exp_rx_pkts =  data_flow_rate_percent / 100.0 * speed_gbps \
+            * 1e9 * data_flow_dur_sec / 8.0 / data_pkt_size
+
+        deviation = (rx_frames - exp_rx_pkts) / float(exp_rx_pkts)
```

Comment:
why do we need a tolerance check for this test?  tx==rx should be good enough?

Reply:
I also add a throughput check to ensure that enabling PFC watchdog does not delay any packets.